### PR TITLE
Prevent dual enrollment 13847

### LINKED ISSUE #13847: Facility > Classes - It's possible to have a user simultaneously as a coach and a learner
❌ **This issue is not open for contribution. Visit <a href="https://learningequality.org/contributing-to-our-open-code-base/" target="_blank">Contributing guidelines</a>** to learn about the contributing process and how to find suitable issues.

## Observed behavior

Noticed that previously at `Facility > Classes` we were not allowing to have a user simultaneously as a coach and a learner because if I had already enrolled a coach user in a class as a learner, then I couldn't assign the same coach as a coach user in the class. Now by using the options provided at the "Users" page it's possible to do so:

https://github.com/user-attachments/assets/74e510f7-2588-4887-8dcb-90484023a225

## Expected behavior

To be further discussed and specified.
 
## Steps to reproduce the issue

1. Install the build from https://github.com/learningequality/kolibri/releases/tag/v0.19.0-beta1
2. Go to Device > Classes and attempt to have a user of type super admin, admin or coach act as both a learner and a coach user.
3. Go to the user's table and attempt to do the same and observe that now the user is added both as a coach and a learner.

## Usage Details
[﻿Kolibri 0.19.0b1](https://github.com/learningequality/kolibri/releases/tag/v0.19.0-beta1)
Windows 11, Ubuntu 22, Mac OS 15.5  - Chrome, Firefox

### PR DESCRIPTION
**Pull request by nucleogenesis**
## Summary

_Work planned interactively with Claude Code (+superpowers),  some (mostly backend) code updates performed by Claude code._

Disallows in the API the ability to enroll a user to a class if they're already assigned to that class as a coach (or vice 
versa).

This maintains the same behavior of enrolling/assigning users through the Class Details page, where we are filtering out the users which are not permitted.

## References

Fixes #13847 

#### Known issues 

- You should not see any error when you try to add a user who would not be added - the snackbar "users enrolled successfully" may not be 100% true, but it's not in scope here as UX writing review can help us poilsh this UX
- **Currently** there is just a message like "Users already in selected classes will not be affected" - we _may_ update some of this through the UX writing reviwe

#### Fixes the related issue

Basically, you should be able to use the bulk user management flows without issue such that if you try to assign/enroll a user who is not permitted to be assigned/enrolled _because they're already one or the other_. Users included in the selection are basically ignored/filtered out during the request to assign/enroll them - so there should be basically no effect on that user's relation to the class(es) selected.

- Test one user who should be ignored
- Test a selection with a mix of users - some who will certainly be enrolled/assigned and others who won't
- Test both of the above with 1 selected class, and when selecting multiple classes to assign/enroll them to
- Be sure to test both directions (ie, an admin enrolled as a learner cannot be then assigned as a coach, and vice versa).

### CODE REVIEW

File: kolibri/core/auth/test/test_api.py [hunk 1]
```diff
@@ -2332,6 +2332,247 @@ def test_delete_does_not_affect_other_user_memberships(self):
...
+        )
+
+        # Verify user2 was enrolled
+        self.assertTrue(
+            models.Membership.objects.filter(
+                user=user2, collection=self.classroom
+            ).exists()
+        )
+
+    def test_can_enroll_in_learnergroup(self):
```
Comment (nucleogenesis) [inline kolibri/core/auth/test/test_api.py:2484]: Not sure this is necessary but Claude seemed concerned during planning about how this could/would impact learnergroups -- but really there isn't a path for this to be an issue in the UI

Comment (nucleogenesis) [reply kolibri/core/auth/test/test_api.py:2484]: I kept it because it's something that should pass so long as we're not making huge changes to the UI and it's probably a good place to see a failure if those changes touch this API anyway

**Review by rtibbles — COMMENTED** [review]
My only thoughts are:

1. Silently 201ing feels slightly icky, but I get it. As it's an edge case, the alternative would be to just raise a 400 and let the frontend try again.
2. For both roles and membership, it's doing an exists query for every item that's being created, so you will have an N queries issue. Might be better to collect user lists by collection and then do a check for each collection and cross referencing the user_id lists. Alternatively, if you're just doing a 400 for the entire request, you could do an exists check for each colelction and raise the ValidationError there and then.
3. Are we ever using the individual model serializer on its own, or are we always handling even a single create/update using the ListSerializer? If not, then this will still be possible one user at a time.

File: kolibri/core/auth/serializers.py [hunk 2]
```diff
@@ -28,6 +28,61 @@
...
+                classroom_coach_items.append(role_data)
+            else:
+                other_items.append(role_data)
+
+        # If no classroom coach items, return everything as-is
+        if not classroom_coach_items:
+            return attrs
+
+        # Group classroom coach items by collection to minimize queries
+        items_by_collection = defaultdict(list)
```
Comment (nucleogenesis) [inline kolibri/core/auth/serializers.py]: This was the tree I started barking up by hand exploring - there's obviously a lot more code here and a lot of iterating over things in general @rtibbles - but it does address the N queries problem I think. As I read through it (thanks Claude) my main negative thought was "seems like a lot of code for what feels like it should be simpler to do".

At a high level, though (and to your point # 1 re: the ickiness of silent 201s), hindsight being 20/20 I think that we might have done well to deliberately make for a partial success API call that includes some info in the response about what did/didn't happen. We never intended to display such granular errors to the user, necessarily, but it could be used to answer certain questions such as "did _anything_ happen at all?" because in this case you could select 2 coaches and try to enroll them in a class they're filtered out of and still we'll say "Users enrolled successfully".

**Review by pcenov — APPROVED** [review]
No issues observed while manually testing - implemented as specified!

**Review by rtibbles — CHANGES_REQUESTED** [review]
It's an edge case (but so is the dual enrollment prevention), but telling a user that the coach has been assigned, when it definitely hasn't, seems pretty misleading (and is a result of our silently swallowing errors).

https://github.com/user-attachments/assets/54e460c9-e7df-4d2d-a355-9953aa804c6a

**Comment by pcenov** [comment]
Hi @rtibbles and @nucleogenesis - yesterday while testing I saw the edge case based on which Richard is requesting changes here, but I thought that this is corresponding to the general idea of this fix. I don't see how this edge case is different from the entirely realistic scenario where if I go to the Users table, select exactly the same coach user and assign that user as a coach (getting a misleading/entirely wrong success message) since the coach user is actually already enrolled as a learner?

**Comment by nucleogenesis** [comment]
@pcenov This PR needs some additional review as we've opted to include some substantial changes that handle edge cases around partial success/failure.

Please test the following scenarios in bulk user management workflow on the users table:

1) When all of your selected users can successfully be enrolled/assigned (ie, they're not assigned/enrolled and *should* succeed)
2) When some, but not all, of your users can successfully be enrolled/assigned (ie, assign some users who are enrolled and some who are not -- test in both directions)
3) When none of your selected users would successfully be enrolled/assigned (ie, assign a selection of users who are enrolled -- test in both directions)


In the "Class Details" page, please test the edge case of multi-tab scenarios where one tab opens to see a list of users who are valid, then make it so they cannot be assigned/enrolled (based on the circumstances)  in another tab, go back and try to assign the users who should no longer be allowed to be assigned/enrolled in that situation.

Generally, you should see correct snackbar messages.

---

Things to expect to see:

1) When there are NO failures at all: the workflow should be basically the same as before - we close the side panel and go back to a unfiltered table.
2) When there are ANY failures:

* the side panel will close
* the snackbar message will indicate the partial success
* the user will see the UsersTable  pre-filtered to only include the users who failed to be assigned/enrolled
* there should also be a correct message alert at the top of the table about the failure
* clearing the filters should result in the alert message disappearing

---

**Comment by pcenov** [comment]
Hi @nucleogenesis,

**Users page:**

1. If I have already assigned a user as a coach, then when I attempt to enroll the same user as a learner I am seeing (only for a split second) a yellow banner **"The users shown were unable to be enrolled"**, and I am seeing the table with **all** available users, plus the following error in the console: `ReferenceError: invalidRoles is not defined`

https://github.com/user-attachments/assets/ac9bd2a1-b599-4892-8c3d-6c93b3e4a798

Similar behavior is observed when attempting to enroll several coaches as learners while some of them are already enrolled as such:

https://github.com/user-attachments/assets/7a7b17df-e305-4630-8ad8-1215d296b2b5

2. If I have already enrolled some of my coaches to a class as learners, then when I attempt to assign all of the coaches to all available classes I am seeing the following snackbar message `Some coaches assigned successfully, but some were not.` (which is correct by itself) and I am seeing only the table rows for the coaches which were not assigned to a class (since they are enrolled as learners) BUT while partially helpful, this can still be very confusing from a user's point of view, as the snackbar message is shown only for 5 seconds and also I am still seeing the text `11 users selected <ins>Clear selection</ins>` while there are now only 3 coaches in the table... Nothing happens when I attempt to click on the `Clear selection` link and I am seeing a bunch of `TypeError: someFailedToAssign$ is not a function` errors in the console.

https://github.com/user-attachments/assets/204e0227-0497-4ce8-8a02-e9ed2623a65e

3. If a user (regardless of the role) is not enrolled in a class, then when I try to enroll that user (or multiple users) in a class, I am always brought back to the page I was before completing the action:

https://github.com/user-attachments/assets/d01e2ce1-d223-48e7-ae83-e189be74f01a

4. At the `New users` page things are working better, but I'm not seeing the yellow banner at all, and when trying to enroll already enrolled learners, the list with users is not filtered:

https://github.com/user-attachments/assets/01017d32-6033-45dd-ab24-8972f11f37d9

**Classes page:**

All looks good to me, except the following potentially confusing edge cases:

1. An edge case where I am at the `Classes` page and I am keeping the `Assign coaches` tab open separately, while at the same time I have already enrolled **all** coaches as learners, results in not showing the snackbar message at all:

https://github.com/user-attachments/assets/7986b1cf-4acc-4465-a80c-853220f15682

Trying the same scenario with some coaches enrolled as learners and others not, results in seeing the following snackbar message `Some coaches assigned successfully, but some were not.`, which while being true is potentially not very helpful in that particular scenario especially if they have multiple coaches and they don't know which were assigned and which were not:

https://github.com/user-attachments/assets/ddafa7be-783b-4d43-b693-8b52f850523f

File: kolibri/core/auth/api.py [hunk 3]
```diff
@@ -786,6 +786,27 @@ class MembershipViewSet(BulkDeleteMixin, BulkCreateMixin, viewsets.ModelViewSet)
...
+        serializer.is_valid(raise_exception=True)
+
+        self.perform_create(serializer)
+
+        response_data = {"created": serializer.data}
+        status_code = status.HTTP_201_CREATED
+
+        if hasattr(serializer, "invalid_items") and serializer.invalid_items:
+            status_code = status.HTTP_207_MULTI_STATUS
+            response_data["invalid"] = [
```
Comment (AlexVelezLl) [inline kolibri/core/auth/api.py:800]: (Nitpick) Initially, it was not evident to me why an `invalid_items` property would be set if the `is_valid` method is setting `raise_exception=True` and in theory it would raise an exception if the data is invalid, but this is true for bulk creations where some items may be invalid but not all of them, and therefore it wouldn't raise the exception. It would be great if we could have some comments around here explaining this.

File: kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue [hunk 4]
```diff
@@ -213,17 +223,40 @@
...
+                  query: {
+                    by_ids: invalidMemberships.value.map(r => r.user).join(','),
+                    failedActionType: InvalidActionTypes.ENROLL,
+                  },
+                }),
+              );
+              emit('clearSelection');
+              return true;
+            } else {
+              return true;
```
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue]: Seems like because of these early returns, there are some regressions now on the enroll learners workflow: When a learner is enrolled in a class, but the loading is never set to false, and the side panel is not closed:

https://github.com/user-attachments/assets/e1c63fa4-c361-4a4d-8f3d-9032149b8754

File: kolibri/plugins/facility/assets/src/composables/useUserManagement.js [hunk 5]
```diff
@@ -37,18 +38,28 @@ export default function useUserManagement({
...
+          }
+        : pickBy({
+            member_of: activeFacilityId,
+            date_joined__gte: dateJoinedGt?.toISOString(),
+            page: page.value,
+            page_size: pageSize.value,
+            search: search.value?.trim() || null,
+            ordering: order.value === 'desc' ? `-${ordering.value}` : ordering.value || null,
+            ...getBackendFilters(),
+          });
```
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/composables/useUserManagement.js]: Just noting that since we are not sending the `ordering` param, the column sorting is not working properly in this view.


https://github.com/user-attachments/assets/db86885a-557e-4e7d-950f-eb2b836ac987

File: kolibri/plugins/facility/assets/src/composables/useUsersFilters.js
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/composables/useUsersFilters.js:1]: Not sure how relevant it is, but just noting that if we open the filters, we can still see the previously applied filters, which may or may not be confusing:

image

File: kolibri/plugins/facility/assets/src/composables/useUsersFilters.js [hunk 6]
```diff
@@ -119,6 +124,13 @@ export default function useUsersFilters({ classes }) {
   watch(
     routeFilters,
     newFilters => {
+      if (workingFilters.by_ids?.length) {
+        // If we're filtering by by_ids, it's programmatic navigation
+        // so we will clear all existing filters and set the ids alone
+        resetWorkingFilters();
+        workingFilters.by_ids = newFilters.by_ids;
+        return;
+      }
```
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/composables/useUsersFilters.js]: It's just an idea, but I feel we are adding too many edgy handlers for the by_ids filters that may be easily overseen if we update these modules in the future. I wonder if it would be better just to explicitly reset all filters when applying the `by_ids` filter, and let this filter co-exist with others without too many edge case handlers.

So, in the side panel modals, we may call the `resetWorkingFilter` and apply filters with just a `by_ids` filter, so this would override the filters on the route query leaving the `by_ids` filter always isolated, but isolated by on purposely removing all other filters from its source of truth and not by adding conditionals to cut the filters execution in the composables.

File: kolibri/plugins/facility/assets/src/modules/classAssignMembers/actions.js [hunk 7]
```diff
@@ -1,25 +1,23 @@
...
 
-export function enrollLearnersInClass(store, { classId, users }) {
-  return MembershipResource.saveCollection({
-    getParams: {
-      collection: classId,
-    },
+export async function enrollLearnersInClass(store, { classId, users }) {
+  return client({
+    url: urls['kolibri:core:membership_list'](),
+    method: 'POST',
```
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/modules/classAssignMembers/actions.js]: Just curious to know what was wrong with the `saveCollection` method? I imagine it wasn't returning the proper response from the backend?

File: kolibri/plugins/facility/assets/src/views/users/NewUsersPage.vue [hunk 8]
```diff
@@ -357,7 +369,26 @@
         fetchClasses();
       });
 
+      // Alerting users about enrollment/assignment errors
+      const alertDismissed = ref(false);
+
+      const showingInvalidUsers = computed(
+        () => !alertDismissed.value && Boolean(route.query.by_ids),
+      );
```
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/views/users/NewUsersPage.vue]: I am not seeing any other assignment to the `asertDismissed` ref, so this is always `false` in the entire component life cycle? Should we also check the `failedActionType` ref here?

File: kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue [hunk 9]
```diff
@@ -213,17 +223,40 @@
...
+              router.push(
+                overrideRoute(route, {
+                  name: getRootRouteName(route),
+                  query: {
+                    by_ids: invalidMemberships.value.map(r => r.user).join(','),
+                    failedActionType: InvalidActionTypes.ENROLL,
+                  },
+                }),
+              );
+              emit('clearSelection');
```
Comment (AlexVelezLl) [inline kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue]: Just noting that the `props.onChange` already takes care of clearing the selection, too. Should we use that instead?

**Review by AlexVelezLl — COMMENTED** [review]
Thanks @nucleogenesis! Just one little regression I found on the enroll learners side panel. Everything else are more nitpicky comments!

File: kolibri/plugins/facility/assets/src/modules/classAssignMembers/actions.js (hunk 7 above)
Comment (rtibbles) [reply kolibri/plugins/facility/assets/src/modules/classAssignMembers/actions.js]: Yes, saveCollection assumes a very specific shape to the returned values, which was being subverted by the endpoint, because of the way it was returning the errored users.

File: kolibri/plugins/facility/assets/src/views/users/NewUsersPage.vue (hunk 8 above)
Comment (nucleogenesis) [reply kolibri/plugins/facility/assets/src/views/users/NewUsersPage.vue]: I thought I got rid of all of that stuff, great catch! Thanks

File: kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue (hunk 9 above)
Comment (nucleogenesis) [reply kolibri/plugins/facility/assets/src/views/users/sidePanels/EnrollLearnersSidePanel.vue]: Thanks - I think the early return was the cause of this unnecessary solution!

File: kolibri/plugins/facility/assets/src/composables/useUserManagement.js (hunk 5 above)
Comment (nucleogenesis) [reply kolibri/plugins/facility/assets/src/composables/useUserManagement.js]: Thank you!!

File: kolibri/plugins/facility/assets/src/composables/useUsersFilters.js (hunk 6 above)
Comment (nucleogenesis) [reply kolibri/plugins/facility/assets/src/composables/useUsersFilters.js]: I need to come back to this comment tomorrow with a fresher mind but I like your suggestion here . I also think that taking pass at what you're talking about here will help me better work out if or how we should handle your comment just above re: filters still being visible

**Comment by pcenov** [comment]
Hi @nucleogenesis, I confirm that the issues I reported are now fixed and that sorting is working when displaying the list of unassigned users.

**Critical regression issue:**

The new user creation flow is broken now when directly enrolling or assigning the newly created user (the user is actually created but the side panel doesn't reflect that fact):

https://github.com/user-attachments/assets/4079cbcc-fd09-4201-863f-4167921523dd

**New issue:**

1. When I am presented with the list of unassigned users, it feels natural to select them, remove them from all classes and then attempt to assign them to all or some of the classes. However in that case the banner persistently remains visible along with the now obsolete list of unassigned users, since they are now assigned:

https://github.com/user-attachments/assets/89ff0e9d-4f93-43ef-a15b-fdd2a3a89fe7

A couple of usability issues that should be further discussed:

1. When I am assigning/enrolling users in multiple classes and some were not assigned/enrolled, then we do show the yellow banner and we do list the users, but it's not clear for which class or classes that applies. So it's probably best to add that info in the banner text otherwise an unsuspecting user might keep trying to assign or enroll the users wondering why it's not working (I guess they should be advised to go to the Classes page and start looking at what's going on):

https://github.com/user-attachments/assets/ebd936e8-7b02-41f3-89af-3513881576a0

Here's an example of me searching for a specific user, trying to assign that user to several classes and not being sure what's failing. Note that the banner text doesn't reflect the fact that we are dealing with a single user, and that the snackbar message text `Some coaches assigned successfully, but some were not.` is really confusing:

https://github.com/user-attachments/assets/d449d9f9-3376-4f45-a6a8-607301906be7

3. When we are presented with the short list of users who were not assigned or enrolled then I can see the `1 filter` link and if I click on it, none of the filters are selected leaving me puzzled. So maybe in that case the `1 filter` link should be hidden.

https://github.com/user-attachments/assets/ddf353b9-67c7-493f-bfa0-27232a2dde2a

Note that it should also be considered that one can have applied some filters prior to assigning/enrolling users, in which case currently if everything is OK we are keeping the filters intact after completing the operation, while if some of the users were not assigned/enrolled we are again showing the `1 filter` link  (while I can clearly see that there are multiple filters applied):

https://github.com/user-attachments/assets/9edd6596-c201-4a3e-a384-8ff7a9b99793

**Comment by pcenov** [comment]
Hi @nucleogenesis,

1. The critical regression is fixed for users of type learner, facility coach and admin, however when I try to create a `class coach` user, select a class and click either the `Save and close` or `Save and add another` button then I am getting a 500 error in the console plus `TypeError: this.errors.includes is not a function`:

https://github.com/user-attachments/assets/0e19b5b3-0f75-4d75-ae83-3c7096cc16da

2. After creating a new user the user is not immediately listed in the New users table, have to manually refresh the page to see the user:

https://github.com/user-attachments/assets/b28fa9ae-c5c1-4aac-b9c5-790f075ca089

3. When a user cannot be assigned then at the `Users` page you are showing the yellow banner, while at the `New users` page you are not:

https://github.com/user-attachments/assets/732ef861-fd41-4512-8101-e11e612aa623

From the list with the usability issues in my previous comment I can see that only the last point has been fixed, so lets not forget to discuss the rest as they are important. :)

**Comment by nucleogenesis** [comment]
@pcenov thank you so much for your thorough review of everything - I've fixed the regressions you noted and this is ready for another look.

**Comment by pcenov** [comment]
@nucleogenesis thank you as well for your tenacious efforts here!

I confirm that point 1 and 3 are fixed now.

Point 2 from above is partially fixed. If I assign/enroll the user to a class while creating it then the user is displayed in the New users table but if I create a user without assigning or enrolling that user, then the user is not immediately displayed in the New users table:

https://github.com/user-attachments/assets/277028b0-d90e-448d-b43a-787c2793b94f

The other regression I observed is for the `Copy class` option - now when I select 'Copy class' I can see that the 'Copy all learners' checkbox is preselected while 'Copy all coaches' is not selected (both are supposed to be unchecked in that case). Selecting either of the checkboxes or both and clicking the 'Make a copy' button always results in the following error:

```
{
  "data": {
    "created": [
      {
        "id": "c9f71524af483ca2e28778a8a242b90c",
        "kind": "coach",
        "collection": "8ef213e15938f5a39f39c33f214bc58e",
        "user": "00a96b143126b7c51f52a0c213a049a2"
      }
    ]
  },
  "status": 201,
  "statusText": "Created",
  "headers": {
    "allow": "GET, POST, DELETE, HEAD, OPTIONS",
    "content-length": "160",
    "content-security-policy": "script-src 'self' blob:; default-src 'self' data: blob:; style-src 'self' data: blob: 'unsafe-inline'; frame-src 'self' data: blob: http: https:",
    "content-type": "application/json",
    "date": "Mon, 03 Nov 2025 13:38:38 GMT",
    "referrer-policy": "same-origin",
    "server": "Cheroot/10.0.1",
    "vary": "Accept, Cookie",
    "x-content-type-options": "nosniff",
    "x-frame-options": "DENY"
  },
  "config": {
    "transitional": {
      "silentJSONParsing": true,
      "forcedJSONParsing": true,
      "clarifyTimeoutError": false
    },
    "adapter": [
      "xhr",
      "http",
      "fetch"
    ],
    "transformRequest": [
      null
    ],
    "transformResponse": [
      null
    ],
    "timeout": 0,
    "xsrfCookieName": "kolibri_csrftoken",
    "xsrfHeaderName": "X-CSRFToken",
    "maxContentLength": -1,
    "maxBodyLength": -1,
    "env": {},
    "headers": {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json",
      "X-Requested-With": "XMLHttpRequest"
    },
    "paramsSerializer": {},
    "url": "/api/auth/role/",
    "data": "[{\"user\":\"00a96b143126b7c51f52a0c213a049a2\",\"kind\":\"coach\",\"collection\":\"8ef213e15938f5a39f39c33f214bc58e\"}]",
    "method": "post",
    "allowAbsoluteUrls": true
  },
  "request": {}
}
```

Here's a video of that:

https://github.com/user-attachments/assets/251e06de-4ecd-43ab-9501-b2cfca8bb4c1

Note that if I close the error window I can see that the class is actually copied successfully. Previously I had reported a similar issue here https://github.com/learningequality/kolibri/issues/13827 but this one is clearly different in nature as it happens always and there's the preselected checkbox regression.

**Comment by marcellamaki** [comment]
> The other regression I observed is for the Copy class option - now when I select 'Copy class' I can see that the 'Copy all learners' checkbox is preselected while 'Copy all coaches' is not selected (both are supposed to be unchecked in that case).

~This is a me problem. Richard and I discussed making this change intentionally (i.e. having the learners selected by default), and I opened a PR but clearly we missed the error that you outline here. I can be responsible for fixing that in follow up, as it is not of Jacob's doing, and probably the result of a rebase. :) ~ 

ETA: discussed w. Richard. the learners selected by default is because of me, but the checkbox is unrelated to the error here, and the reported error can be fixed in the scope of Jacob's PR. 🎢   🙃

**Comment by pcenov** [comment]
Hi @nucleogenesis,

I'm happy to let you know that I was able to identify just the following minor edge case at the `Users` table:

1. The message 'The users shown were unable to be assigned.' doesn't disappear when I decide to instead enroll the users which I was previously unable to assign:

https://github.com/user-attachments/assets/6fcc3500-110b-4a4a-820e-e7a5fc1d80ca

This scenario is working correctly with the other message ("The users shown were unable to be enrolled.").

Everything else seems to be working properly.

At the `Classes` page I noticed the following regression - when I remove some or all of the coaches and try to copy the class then in the modal the count of the coaches is not updated correctly and that results in an error when copying the class:

https://github.com/user-attachments/assets/41c5e7ba-4f9c-4165-80f7-13137d16c14c

File: kolibri/plugins/facility/assets/src/views/LearnerClassEnrollmentPage.vue [hunk 10]
```diff
@@ -72,7 +78,18 @@
...
-          .then(() => {
+          .then(response => {
+            const { created, invalid } = response.data;
+            if (created?.length) {
+              if (invalid?.length) {
+                this.createSnackbar(this.someLearnersEnrolledNotice$());
+              } else {
+                this.createSnackbar(this.usersEnrolledNotice$());
+              }
+              if (invalid?.length) {
```
Comment (rtibbles) [inline kolibri/plugins/facility/assets/src/views/LearnerClassEnrollmentPage.vue]: Is this meant to be an else if with the created?.length check? Otherwise, this is just overriding the someLearnersEnrolledNotice above with noLearnersEnrolled.

File: kolibri/plugins/facility/assets/src/views/CoachClassAssignmentPage.vue [hunk 11]
```diff
@@ -68,13 +74,19 @@
...
+          .then(response => {
+            const { created, invalid } = response.data;
+            if (created?.length) {
+              if (invalid?.length) {
+                this.createSnackbar(this.someCoachesAssignedNotice$());
+              } else {
+                this.createSnackbar(this.coachesAllAssignedNotice$());
+              }
+            }
+            if (invalid?.length) {
```
Comment (rtibbles) [inline kolibri/plugins/facility/assets/src/views/CoachClassAssignmentPage.vue]: Again, shouldn't this be an else if?

File: kolibri/core/auth/serializers.py [hunk 12]
```diff
@@ -28,13 +30,63 @@
...
+        # Tracking the collections & users to query against for validating
+        # if they're members of the class already
+        assignable_coach_ids = []
+        class_collection_ids = []
+
+        for role_data in attrs:
+            collection = role_data["collection"]
+            kind = role_data["kind"]
+
+            # Only validate classroom-level coach roles
```
Comment (rtibbles) [inline kolibri/core/auth/serializers.py:55]: I am a bit confused by this - why are we only preventing classroom assignable coaches from being assigned and enrolled in a class, but not putting similar restrictions on facility coaches and admins?

File: kolibri/core/auth/serializers.py [hunk 13]
```diff
@@ -141,13 +193,53 @@ def validate(self, attrs):
...
+                other_items.append(membership_data)
+
+        # If no classroom items, return everything as-is
+        if not classroom_items:
+            return attrs
+
+        existing_roles = Role.objects.filter(
+            collection_id__in=class_collection_ids,
+            user_id__in=user_ids_to_validate,
+            kind__in=[role_kinds.ASSIGNABLE_COACH, role_kinds.COACH],
```
Comment (rtibbles) [inline kolibri/core/auth/serializers.py]: A user cannot have an "assignable_coach" role wrt a classroom, only wrt a facility.

**Review by rtibbles — CHANGES_REQUESTED** [review]
There are some significant bugs hiding here - obfuscated by an extra wrapper of verbose claude generated code in some cases.

File: kolibri/plugins/facility/assets/src/views/LearnerClassEnrollmentPage.vue (hunk 10 above)
Comment (nucleogenesis) [reply kolibri/plugins/facility/assets/src/views/LearnerClassEnrollmentPage.vue]: This logic made more sense when I was returning early in the side panels like this where no else-if is needed to avoid double-snackbarring

```
      const snackbarMessage$ = () => {
        if (createdRoles.value?.length) {
          if (invalidRoles.value?.length) {
            return someCoachesAssignedNotice$();
          } else {
            return coachesAllAssignedNotice$();
          }
        }
        if (invalidRoles.value?.length) {
          return someFailedToAssign$();
        }
      };
```

Will push updates here shortly

Comment (rtibbles) [reply kolibri/plugins/facility/assets/src/views/LearnerClassEnrollmentPage.vue]: Yeah, I saw the other instances where this was a return early pattern instead, and saw where this came from.

**Comment by nucleogenesis** [comment]
Closing this PR - anything we need from it we can cherry-pick separately
